docs/reference/glossary.md

```diff
@@ -181,22 +187,62 @@ labeled using [tags](#tag).
 Here is an example of the shared [nginx repository](https://hub.docker.com/_/nginx/)
 and its [tags](https://hub.docker.com/r/library/nginx/tags/)
 
-## Swarm
 
-[Swarm](https://github.com/docker/swarm) is a native clustering tool for Docker.
-Swarm pools together several Docker hosts and exposes them as a single virtual
-Docker host. It serves the standard Docker API, so any tool that already works
-with Docker can now transparently scale up to multiple hosts.
+## service
+
+A [service](../swarm/how-swarm-mode-works/services.md) is the definition of how
+you want to run your application containers in a swarm. At the most basic level
+a service  defines which container image to run in the swarm and which commands
+to run in the container. For orchestration purposes, the service defines the
+"desired state", meaning how many containers to run as tasks and where to deploy
+the containers.
+
+Frequently a service is a microservice within the context of some larger
+application. Examples of services might include an HTTP server, a database, or
+any other type of executable program that you wish to run in a distributed
+environment.
+
+## service discovery
+
+Swarm mode [service discovery](../swarm/networking.md) is a DNS component
+internal to the swarm that automatically assigns each service on an overlay
+network in the swarm a VIP and DNS entry. Containers on the network share DNS
+mappings for the service via gossip so any container on the network can access
+the service via its service name.
+
+You don’t need to expose service-specific ports to make the service available to other services on the same overlay network. The swarm’s internal load balancer automatically distributes requests to the service VIP among the active tasks.
+
+## swarm
+
+A [swarm](../swarm/index.md) is a cluster of one or more Docker Engines running in swarm mode.
+
+[Docker Swarm](https://github.com/docker/swarm) is a native clustering tool for
+Docker. Swarm pools together several Docker hosts and exposes them as a single
+virtual Docker host. It serves the standard Docker API, so any tool that already
+works with Docker can now transparently scale up to multiple hosts.
 
 *Also known as : docker-swarm*
 
+## swarm mode
+
+[Swarm mode](../swarm/index.md) refers to cluster management and orchestration
+features embedded in Docker Engine. When you initialize a new swarm (cluster) or
+join nodes to a swarm, the Docker Engine runs in swarm mode.
+
 ## tag
 
 A tag is a label applied to a Docker image in a [repository](#repository).
 tags are how various images in a repository are distinguished from each other.
 
 *Note : This label is not related to the key=value labels set for docker daemon*
 
+## task
+
+A [task](../swarm/how-swarm-mode-works/services.md#tasks-and-scheduling) is the
+atomic unit of scheduling within a swarm. A task carries a Docker container and
```

Comment:
not sure if we should prepare for that already, but "tasks" _currently_ carries containers, but are generic by design (e.g. in future, a task could make coffee, or create/handle volume tasks) - ask Stephen/Andrea if I have understood that correctly.